### ISSUE TITLE
IOExeption, sleep and attempt a reconnect

### ISSUE THREAD

**Issue by anderbubble** [irods/irods_client_nfsrods#95]
Our users are complaining about generally unreliable experiences using our NFSRODS deployment, so I am auditing our logs for error messages. Among them we have 753,863 instances of the error "ERROR Thread-{num} [IRODSBasicTCPConnection] - IOExeption, sleep and attempt a reconnect".

A sample of 10000 instances indicates that this is the most common traceback:

```
java.net.ConnectException: Connection refused (Connection refused)
	at java.net.PlainSocketImpl.socketConnect(Native Method) ~[?:1.8.0_222]
	at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350) ~[?:1.8.0_222]
	at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206) ~[?:1.8.0_222]
	at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188) ~[?:1.8.0_222]
	at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392) ~[?:1.8.0_222]
	at java.net.Socket.connect(Socket.java:589) ~[?:1.8.0_222]
	at java.net.Socket.connect(Socket.java:538) ~[?:1.8.0_222]
	at org.irods.jargon.core.connection.IRODSBasicTCPConnection.connect(IRODSBasicTCPConnection.java:142) ~[nfsrods-0.9.3-jar-with-dependencies.jar:?]
	at org.irods.jargon.core.connection.AbstractConnection.initializeConnection(AbstractConnection.java:232) ~[nfsrods-0.9.3-jar-with-dependencies.jar:?]
	at org.irods.jargon.core.connection.AbstractConnection.<init>(AbstractConnection.java:197) ~[nfsrods-0.9.3-jar-with-dependencies.jar:?]
	at org.irods.jargon.core.connection.IRODSBasicTCPConnection.<init>(IRODSBasicTCPConnection.java:51) ~[nfsrods-0.9.3-jar-with-dependencies.jar:?]
	at org.irods.jargon.core.connection.IRODSTCPConnectionFactoryImpl.instance(IRODSTCPConnectionFactoryImpl.java:33) ~[nfsrods-0.9.3-jar-with-dependencies.jar:?]
	at org.irods.jargon.core.connection.AbstractIRODSMidLevelProtocolFactory.instance(AbstractIRODSMidLevelProtocolFactory.java:86) ~[nfsrods-0.9.3-jar-with-dependencies.jar:?]
	at org.irods.jargon.core.connection.IRODSProtocolManager.createNewProtocol(IRODSProtocolManager.java:139) ~[nfsrods-0.9.3-jar-with-dependencies.jar:?]
	at org.irods.jargon.core.connection.IRODSSimpleProtocolManager.getIRODSProtocol(IRODSSimpleProtocolManager.java:56) ~[nfsrods-0.9.3-jar-with-dependencies.jar:?]
	at org.irods.jargon.core.connection.IRODSSession.connectAndAddToProtocolsMap(IRODSSession.java:487) ~[nfsrods-0.9.3-jar-with-dependencies.jar:?]
	at org.irods.jargon.core.connection.IRODSSession.currentConnection(IRODSSession.java:394) ~[nfsrods-0.9.3-jar-with-dependencies.jar:?]
	at org.irods.jargon.core.pub.IRODSGenericAO.<init>(IRODSGenericAO.java:61) ~[nfsrods-0.9.3-jar-with-dependencies.jar:?]
	at org.irods.jargon.core.pub.IRODSGenQueryExecutorImpl.<init>(IRODSGenQueryExecutorImpl.java:57) ~[nfsrods-0.9.3-jar-with-dependencies.jar:?]
	at org.irods.jargon.core.pub.IRODSAccessObjectFactoryImpl.getIRODSGenQueryExecutor(IRODSAccessObjectFactoryImpl.java:281) ~[nfsrods-0.9.3-jar-with-dependencies.jar:?]
	at org.irods.nfsrods.vfs.ReadWriteAclWhitelist.lambda$new$1(ReadWriteAclWhitelist.java:64) ~[nfsrods-0.9.3-jar-with-dependencies.jar:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) ~[?:1.8.0_222]
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308) ~[?:1.8.0_222]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180) ~[?:1.8.0_222]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294) ~[?:1.8.0_222]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) ~[?:1.8.0_222]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) ~[?:1.8.0_222]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_222]
```

But there are other variants.

**Comment by michael-conway** [comment]
@korydraughn this looks like a network issue, not an NFSRods issue, really it would be an underlying Jargon issue anyhow. I'd be curious to see the info/debug logging but it's just trying to connect and not finding the endpoint.

**Comment by korydraughn** [comment]
@michael-conway I agree.

@anderbubble  Can you try increasing the logging and see what else shows up.  NFSRODS does not contain any network code.  It only speaks to Jargon and nfs4j.

Jargon log level:  https://github.com/irods/irods_client_nfsrods/blob/5aecafa369a2f4bcf56ff68707817d9a7ec96aa0/irods-vfs-impl/config/log4j.properties#L19
Root logger log level:  https://github.com/irods/irods_client_nfsrods/blob/5aecafa369a2f4bcf56ff68707817d9a7ec96aa0/irods-vfs-impl/config/log4j.properties#L24

Also, How long has this instance been running?  When did these errors begin to appear?

**Comment by korydraughn** [comment]
Could this be happening because the resources on the iRODS server are exhausted?

**Comment by trel** [comment]
Any news/insight?    Hoping for some network shenanigans...

**Comment by michael-conway** [comment]
Is this a long long running stale connection issue? That would imply a
non-closed conn on an operation? Do we generally see any remaining hung
agents running NFS? That could be a symptom of a place conns are not being
cleaned up.

**Comment by trel** [comment]
NFSRODS now lets go of the connections in a reasonable manner.

Haven't seen any issues since this commit https://github.com/irods/irods_client_nfsrods/commit/f41095777a19b87a9b3b0c8507cba645269764b1  (included in 0.9.3).

**Comment by trel** [comment]
closing - please make a new issue if seen again.
